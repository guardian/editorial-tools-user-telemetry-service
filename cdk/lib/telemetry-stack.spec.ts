import "@aws-cdk/assert/jest";
import { App } from "aws-cdk-lib";
import { TelemetryStack } from "./telemetry-stack";
import { Template } from "aws-cdk-lib/assertions";

describe("The telemetry stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new TelemetryStack(app, "telemetry", {
      stack: "flexible",
      stage: 'TEST',
    });

    const template = Template.fromStack(stack);

    expect(template.toJSON()).toMatchSnapshot();
  });
});
