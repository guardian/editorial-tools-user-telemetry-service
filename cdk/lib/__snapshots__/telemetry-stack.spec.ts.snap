// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The telemetry stack matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "usertelemetryEndpointE4D637B1": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "usertelemetry6D0235CF",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "usertelemetryDeploymentStageprodD71A8078",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BucketName": Object {
      "Description": "Name of the bucket to persist event data",
      "Type": "String",
    },
    "Cert": Object {
      "Description": "Certificate ARN for telemetry endpoint",
      "Type": "String",
    },
    "Hostname": Object {
      "Description": "Hostname for telemetry endpoint",
      "Type": "String",
    },
    "KinesisArn": Object {
      "Description": "ARN of the Kinesis stream to post event data",
      "Type": "String",
    },
    "MaxLogSize": Object {
      "Description": "Maximum size (in bytes) of log data from an individual request",
      "Type": "String",
    },
    "PandaSettingsKey": Object {
      "Description": "The location of the pan-domain authentication settings file",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": Object {
      "DependsOn": Array [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "import boto3  # type: ignore
import json
import logging
import urllib.request

s3 = boto3.client(\\"s3\\")

EVENTBRIDGE_CONFIGURATION = 'EventBridgeConfiguration'

CONFIGURATION_TYPES = [\\"TopicConfigurations\\", \\"QueueConfigurations\\", \\"LambdaFunctionConfigurations\\"]

def handler(event: dict, context):
  response_status = \\"SUCCESS\\"
  error_message = \\"\\"
  try:
    props = event[\\"ResourceProperties\\"]
    bucket = props[\\"BucketName\\"]
    notification_configuration = props[\\"NotificationConfiguration\\"]
    request_type = event[\\"RequestType\\"]
    managed = props.get('Managed', 'true').lower() == 'true'
    stack_id = event['StackId']

    if managed:
      config = handle_managed(request_type, notification_configuration)
    else:
      config = handle_unmanaged(bucket, stack_id, request_type, notification_configuration)

    put_bucket_notification_configuration(bucket, config)
  except Exception as e:
    logging.exception(\\"Failed to put bucket notification configuration\\")
    response_status = \\"FAILED\\"
    error_message = f\\"Error: {str(e)}. \\"
  finally:
    submit_response(event, context, response_status, error_message)

def handle_managed(request_type, notification_configuration):
  if request_type == 'Delete':
    return {}
  return notification_configuration

def handle_unmanaged(bucket, stack_id, request_type, notification_configuration):
  external_notifications = find_external_notifications(bucket, stack_id)

  if request_type == 'Delete':
    return external_notifications

  def with_id(notification):
    notification['Id'] = f\\"{stack_id}-{hash(json.dumps(notification, sort_keys=True))}\\"
    return notification

  notifications = {}
  for t in CONFIGURATION_TYPES:
    external = external_notifications.get(t, [])
    incoming = [with_id(n) for n in notification_configuration.get(t, [])]
    notifications[t] = external + incoming

  if EVENTBRIDGE_CONFIGURATION in notification_configuration:
    notifications[EVENTBRIDGE_CONFIGURATION] = notification_configuration[EVENTBRIDGE_CONFIGURATION]
  elif EVENTBRIDGE_CONFIGURATION in external_notifications:
    notifications[EVENTBRIDGE_CONFIGURATION] = external_notifications[EVENTBRIDGE_CONFIGURATION]

  return notifications

def find_external_notifications(bucket, stack_id):
  existing_notifications = get_bucket_notification_configuration(bucket)
  external_notifications = {}
  for t in CONFIGURATION_TYPES:
    external_notifications[t] = [n for n in existing_notifications.get(t, []) if not n['Id'].startswith(f\\"{stack_id}-\\")]

  if EVENTBRIDGE_CONFIGURATION in existing_notifications:
    external_notifications[EVENTBRIDGE_CONFIGURATION] = existing_notifications[EVENTBRIDGE_CONFIGURATION]

  return external_notifications

def get_bucket_notification_configuration(bucket):
  return s3.get_bucket_notification_configuration(Bucket=bucket)

def put_bucket_notification_configuration(bucket, notification_configuration):
  s3.put_bucket_notification_configuration(Bucket=bucket, NotificationConfiguration=notification_configuration)

def submit_response(event: dict, context, response_status: str, error_message: str):
  response_body = json.dumps(
    {
      \\"Status\\": response_status,
      \\"Reason\\": f\\"{error_message}See the details in CloudWatch Log Stream: {context.log_stream_name}\\",
      \\"PhysicalResourceId\\": event.get(\\"PhysicalResourceId\\") or event[\\"LogicalResourceId\\"],
      \\"StackId\\": event[\\"StackId\\"],
      \\"RequestId\\": event[\\"RequestId\\"],
      \\"LogicalResourceId\\": event[\\"LogicalResourceId\\"],
      \\"NoEcho\\": False,
    }
  ).encode(\\"utf-8\\")
  headers = {\\"content-type\\": \\"\\", \\"content-length\\": str(len(response_body))}
  try:
    req = urllib.request.Request(url=event[\\"ResponseURL\\"], headers=headers, data=response_body, method=\\"PUT\\")
    with urllib.request.urlopen(req) as response:
      print(response.read().decode(\\"utf-8\\"))
    print(\\"Status code: \\" + response.reason)
  except Exception as e:
      print(\\"send(..) failed executing request.urlopen(..): \\" + str(e))
",
        },
        "Description": "AWS CloudFormation handler for \\"Custom::S3BucketNotifications\\" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "python3.7",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "s3:GetBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": Array [
          Object {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventApiLambda6AA4ADA7": Object {
      "DependsOn": Array [
        "EventApiLambdaServiceRoleDefaultPolicy92CD0566",
        "EventApiLambdaServiceRole3695319B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "composer-dist",
          "S3Key": "flexible/TEST/event-api-lambda/event-api-lambda.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "user-telemetry",
            "LOG_ENDPOINT_ENABLED": "true",
            "MAX_LOG_SIZE": Object {
              "Ref": "MaxLogSize",
            },
            "PANDA_SETTINGS_KEY": Object {
              "Ref": "PandaSettingsKey",
            },
            "STACK": "flexible",
            "STAGE": "TEST",
            "TELEMETRY_BUCKET_NAME": Object {
              "Ref": "BucketName",
            },
          },
        },
        "FunctionName": "event-api-lambda-TEST",
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "EventApiLambdaServiceRole3695319B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "user-telemetry",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EventApiLambdaServiceRole3695319B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "user-telemetry",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventApiLambdaServiceRoleDefaultPolicy92CD0566": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "BucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "BucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventApiLambdaServiceRoleDefaultPolicy92CD0566",
        "Roles": Array [
          Object {
            "Ref": "EventApiLambdaServiceRole3695319B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventS3LambdaA2D7D547": Object {
      "DependsOn": Array [
        "EventS3LambdaServiceRoleDefaultPolicyC5FCC514",
        "EventS3LambdaServiceRole34104ADE",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "composer-dist",
          "S3Key": "flexible/TEST/event-s3-lambda/event-s3-lambda.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "user-telemetry",
            "LOG_ENDPOINT_ENABLED": "true",
            "MAX_LOG_SIZE": Object {
              "Ref": "MaxLogSize",
            },
            "STACK": "flexible",
            "STAGE": "TEST",
            "TELEMETRY_BUCKET_NAME": Object {
              "Ref": "BucketName",
            },
            "TELEMETRY_STREAM_NAME": Object {
              "Fn::Select": Array [
                1,
                Object {
                  "Fn::Split": Array [
                    "/",
                    Object {
                      "Fn::Select": Array [
                        5,
                        Object {
                          "Fn::Split": Array [
                            ":",
                            Object {
                              "Ref": "KinesisArn",
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          },
        },
        "FunctionName": "event-s3-lambda-TEST",
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "EventS3LambdaServiceRole34104ADE",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "user-telemetry",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EventS3LambdaServiceRole34104ADE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "user-telemetry",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventS3LambdaServiceRoleDefaultPolicyC5FCC514": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "BucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "BucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "kinesis:PutRecords",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "KinesisArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventS3LambdaServiceRoleDefaultPolicyC5FCC514",
        "Roles": Array [
          Object {
            "Ref": "EventS3LambdaServiceRole34104ADE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "telemetrybucketAllowBucketNotificationsTotelemetryEventS3LambdaF393B67EFDEA6CDB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EventS3LambdaA2D7D547",
            "Arn",
          ],
        },
        "Principal": "s3.amazonaws.com",
        "SourceAccount": Object {
          "Ref": "AWS::AccountId",
        },
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":s3:::",
              Object {
                "Ref": "BucketName",
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "telemetrybucketNotifications3F9E1150": Object {
      "DependsOn": Array [
        "telemetrybucketAllowBucketNotificationsTotelemetryEventS3LambdaF393B67EFDEA6CDB",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "BucketName",
        },
        "Managed": false,
        "NotificationConfiguration": Object {
          "LambdaFunctionConfigurations": Array [
            Object {
              "Events": Array [
                "s3:ObjectCreated:*",
              ],
              "LambdaFunctionArn": Object {
                "Fn::GetAtt": Array [
                  "EventS3LambdaA2D7D547",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "telemetrydnsrecordTEST": Object {
      "Properties": Object {
        "Name": Object {
          "Ref": "Hostname",
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "usertelemetrydomainnameTESTF3EFC436",
              "DistributionDomainName",
            ],
          },
        ],
        "Stage": "TEST",
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "usertelemetry6D0235CF": Object {
      "Properties": Object {
        "EndpointConfiguration": Object {
          "Types": Array [
            "EDGE",
          ],
        },
        "Name": "user-telemetry",
        "Policy": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "usertelemetryANY08DB1886": Object {
      "Properties": Object {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "EventApiLambda6AA4ADA7",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "usertelemetry6D0235CF",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "usertelemetryANYApiPermissionTesttelemetryusertelemetry6ED2ABD8ANY598D9524": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EventApiLambda6AA4ADA7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "usertelemetry6D0235CF",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "usertelemetryANYApiPermissiontelemetryusertelemetry6ED2ABD8ANYD968F76A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EventApiLambda6AA4ADA7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "usertelemetry6D0235CF",
              },
              "/",
              Object {
                "Ref": "usertelemetryDeploymentStageprodD71A8078",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "usertelemetryAccount5FDE92D4": Object {
      "DependsOn": Array [
        "usertelemetry6D0235CF",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "usertelemetryCloudWatchRole308CB7B7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "usertelemetryCloudWatchRole308CB7B7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "usertelemetryDeployment7E6104C110775ae7bde40fffbc3476e0df12b7be": Object {
      "DependsOn": Array [
        "usertelemetryproxyANY639725AB",
        "usertelemetryproxyEC8A2D95",
        "usertelemetryANY08DB1886",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "usertelemetryDeploymentStageprodD71A8078": Object {
      "DependsOn": Array [
        "usertelemetryAccount5FDE92D4",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "usertelemetryDeployment7E6104C110775ae7bde40fffbc3476e0df12b7be",
        },
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
        "StageName": "prod",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "usertelemetrydomainnameTESTF3EFC436": Object {
      "Properties": Object {
        "CertificateArn": Object {
          "Ref": "Cert",
        },
        "DomainName": Object {
          "Ref": "Hostname",
        },
        "EndpointConfiguration": Object {
          "Types": Array [
            "EDGE",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "45.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/editorial-tools-user-telemetry-service",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "usertelemetrydomainnameTESTMaptelemetryusertelemetry6ED2ABD89A1E2F1F": Object {
      "Properties": Object {
        "BasePath": "",
        "DomainName": Object {
          "Ref": "usertelemetrydomainnameTESTF3EFC436",
        },
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
        "Stage": Object {
          "Ref": "usertelemetryDeploymentStageprodD71A8078",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "usertelemetryproxyANY639725AB": Object {
      "Properties": Object {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "EventApiLambda6AA4ADA7",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "usertelemetryproxyEC8A2D95",
        },
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "usertelemetryproxyANYApiPermissionTesttelemetryusertelemetry6ED2ABD8ANYproxy25554F9C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EventApiLambda6AA4ADA7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "usertelemetry6D0235CF",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "usertelemetryproxyANYApiPermissiontelemetryusertelemetry6ED2ABD8ANYproxy795F93FB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EventApiLambda6AA4ADA7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "usertelemetry6D0235CF",
              },
              "/",
              Object {
                "Ref": "usertelemetryDeploymentStageprodD71A8078",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "usertelemetryproxyEC8A2D95": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "usertelemetry6D0235CF",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "usertelemetry6D0235CF",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
  },
}
`;
